Vulkan support

C4e10VeK opened the pull request.

C4e10VeK committed:
refact graphics folder and added graphics config

C4e10VeK committed:
added shader interface and graphics config

Kelvium commented on src/client/graphics/OpenGL/Shader.h:
```diff
@@ -1,10 +1,10 @@
-#include <glm/gtc/matrix_transform.hpp>
-#include <glm/gtc/type_ptr.hpp>
+#include "../Graphics.h"
+#include "glm/glm.hpp"
```
why single quotes?

Kelvium commented on src/client/graphics/OpenGL/SpriteBatch.h:
```diff
@@ -1,15 +1,15 @@
 #define RPG_SPRITEBATCH_H
 
-#include <glm/glm.hpp>
+#include "glm/glm.hpp"
```
why single quotes +1

Kelvium commented on src/client/graphics/OpenGL/Texture.cpp:
```diff
@@ -1,11 +1,11 @@
-#include "Graphics.h"
-#include <stb_image.h>
+#include "../Graphics.h"
+#include "stb_image.h"
```
why single quotes +2

Kelvium commented on src/client/Engine.cpp:
```diff
@@ -2,9 +2,16 @@
+#include "graphics/GraphicsConfig.h"
 
-IWindow &Engine::getWindow(int width, int height, const std::string &title)
+IWindow &Engine::getWindow(const std::string &title)
```
if we use a config for width and height, then why don't we use it for title too?

C4e10VeK replied:
> if we use a config for width and height, then why don't we use it for title too?

I don't think it makes sense to allow the user to change the name of the window via config. The name of the window should have a strictly defined name of the game

Kelvium replied:
> @Kelvium
> 
> > if we use a config for width and height, then why don't we use it for title too?
> 
> I don't think it makes sense to allow the user to change the name of the window via config. The name of the window should have a strictly defined name of the game

that makes sense. but i think setting the window size in config is not convenient. i'd prefer to leave a "TODO: provide a convenient way of changing window size" where the window is created as this is not important and we can implement it when we got to the point where we can make "non-important" tagged bugs/things/etc. :)

C4e10VeK replied:
> > @Kelvium
> > > if we use a config for width and height, then why don't we use it for title too?
> > 
> > 
> > I don't think it makes sense to allow the user to change the name of the window via config. The name of the window should have a strictly defined name of the game
> 
> that makes sense. but i think setting the window size in config is not convenient. i'd prefer to leave a "TODO: provide a convenient way of changing window size" where the window is created as this is not important and we can implement it when we got to the point where we can make "non-important" tagged bugs/things/etc. :)

I thought about it. And I think that the size of the window should be left in the config to save the settings. I thought that the config file will be used to save settings when the user changes something in the game settings

Commit:
Include pch.h everywhere

C4e10VeK committed:
Merge pull request #111 from Kelvium/vulkan-support-fix-pch-included-everywhere

Include pch.h everywhere

C4e10VeK committed:
fix: changed single quotes

Kelvium commented on src/client/graphics/OpenGL/Shader.h:
```diff
@@ -2,7 +2,7 @@
 #include "../Graphics.h"
-#include "glm/glm.hpp"
+#include <glm/glm.hpp>
 #include "glm/gtc/matrix_transform.hpp"
```
here's too

C4e10VeK committed:
changed single quotes

C4e10VeK committed:
change shader to IShader

C4e10VeK committed:
Merge branch 'main' of https://github.com/TrueRPG/TrueRPG into vulkan-support

C4e10VeK committed:
added vk instance creation

C4e10VeK committed:
added logger

C4e10VeK committed:
added debug messenger

C4e10VeK committed:
added destroy

C4e10VeK committed:
fix pch.h include

C4e10VeK commented:
In future we will need to add vulkan to the action

C4e10VeK commented:
I don't understand how working with pch.h

Kelvium commented:
> I don't understand how working with pch.h

Just include pch.h in every .cpp file and the action will pass. This improves build speed

C4e10VeK commented:
> > I don't understand how working with pch.h
> 
> Just include pch.h in every .cpp file and the action will pass. This improves build speed.
> 
> You can see the list of files without pch.h included in the action's log

Thanks

C4e10VeK committed:
fix pch.h include

C4e10VeK committed:
fix pch.h include

C4e10VeK committed:
added debug check for logger

C4e10VeK committed:
added surface creation

C4e10VeK committed:
added logical device creation

C4e10VeK committed:
change Texture to ITexture

C4e10VeK committed:
added vulkan memory allocator

C4e10VeK committed:
fix include pch.h

C4e10VeK committed:
added ISpriteBatch.h

C4e10VeK committed:
update shaders in light systems

C4e10VeK committed:
update sprite batch creation

Commit:
ci: Prepare Vulkan SDK

It downloads Vulkan headers, Vulkan loader and glslangValidator.
It also caches the binaries, so it does not need to redownload
them again.

Commit:
vma: link to Vulkan

It resulted in crash on CI because Vulkan headers were only included
with the Vulkan target.

Also, I included vma right after Vulkan, so it can use the Vulkan
target.

C4e10VeK committed:
Merge pull request #113 from TrueRPG/k/ci-vulkan

ci: Prepare Vulkan SDK

C4e10VeK committed:
added glslang for runtime shade commpilation
